source/common/network/connection_impl.cc

```diff
@@ -780,5 +780,1079 @@ void ClientConnectionImpl::connect() {
+          [this]() -> void { this->onWriteBufferHighWatermark(); },
+          []() -> void { /* TODO(adisuissa): Handle overflow watermark */ })),
+      write_buffer_above_high_watermark_(false), detect_early_close_(true),
+      enable_half_close_(false), read_end_stream_raised_(false), read_end_stream_(false),
+      write_end_stream_(false), current_write_end_stream_(false), dispatch_buffered_data_(false),
+      remote_address_(remote_address), source_address_(source_address),
+      io_timer_(dispatcher.createTimer([this]() { onFileEvent(); })) {
+
+  connecting_ = true;
+  transport_socket_->setTransportSocketCallbacks(*this);
+}
+
+ClientPipeImpl::~ClientPipeImpl() {
+  ASSERT(!isOpen() && delayed_close_timer_ == nullptr,
+         "ClientPipeImpl was unexpectedly torn down without being closed.");
+
+  // In general we assume that owning code has called close() previously to the destructor being
+  // run. This generally must be done so that callbacks run in the correct context (vs. deferred
+  // deletion). Hence the assert above. However, call close() here just to be completely sure that
+  // the fd is closed and make it more likely that we crash from a bad close callback.
+  close(ConnectionCloseType::NoFlush);
+}
+
+void ClientPipeImpl::addWriteFilter(WriteFilterSharedPtr filter) {
+  filter_manager_.addWriteFilter(filter);
+}
+
+void ClientPipeImpl::addFilter(FilterSharedPtr filter) { filter_manager_.addFilter(filter); }
+
+void ClientPipeImpl::addReadFilter(ReadFilterSharedPtr filter) {
+  filter_manager_.addReadFilter(filter);
+}
+
+bool ClientPipeImpl::initializeReadFilters() { return filter_manager_.initializeReadFilters(); }
+
+void ClientPipeImpl::close(ConnectionCloseType type) {
+  if (!isOpen()) {
+    return;
+  }
+
+  uint64_t data_to_write = write_buffer_->length();
+  ENVOY_CONN_LOG(debug, "closing data_to_write={} type={}", *this, data_to_write, enumToInt(type));
+  const bool delayed_close_timeout_set = delayed_close_timeout_.count() > 0;
+  if (data_to_write == 0 || type == ConnectionCloseType::NoFlush ||
+      !transport_socket_->canFlushClose()) {
+    if (data_to_write > 0) {
+      // We aren't going to wait to flush, but try to write as much as we can if there is pending
+      // data.
+      transport_socket_->doWrite(*write_buffer_, true);
+    }
+
+    if (type == ConnectionCloseType::FlushWriteAndDelay && delayed_close_timeout_set) {
+      // The socket is being closed and either there is no more data to write or the data can not be
+      // flushed (!transport_socket_->canFlushClose()). Since a delayed close has been requested,
+      // start the delayed close timer if it hasn't been done already by a previous close().
+      // NOTE: Even though the delayed_close_state_ is being set to CloseAfterFlushAndWait, since
+      // a write event is not being registered for the socket, this logic is simply setting the
+      // timer and waiting for it to trigger to close the socket.
```

Comment:
I think we're going to want to reduce copy and paste between this and the network class.